### ISSUE TITLE
Attributes: Discover without loading the class/file? ("static reflection")

### ISSUE THREAD

**Issue by donquixote** [spiral/framework#460]
## Motivation
One feature of some other annotation discovery tools is to do the discovery without actually loading the source files.
Perhaps we might even want to delay the loading of the actual annotation or attribute classes.

One benefit of this is to "survive" classes that depend on optional 3rd party dependencies, without fatal error.
E.g. if one of your packages has a class that implements a 3rd party interface, but that interface belongs to an optional package that is currently not present, then loading that file would cause a 'Fatal error: Uncaught Error: Interface "I" not found'. See https://3v4l.org/ANDlX. A solution based on a "static reflection parser" can determine that the base class or interface is missing and safely skip the result.

Another benefit might be memory use for php opcode, although I am not really sure if that is a valid concern. Surely the parsing also consumes memory, just a different type of it. And for performance, I would think anything that uses a userland parser must be slower than native loading.

## Current situation in spiral/attributes
The metadata reader in spiral/attributes requires reflection classes.
So we could use the adapter layer from `roave/BetterReflection` to get "fake" reflection classes and functions, and send those to the spiral reader.

But would this not mean that the php file has to be parsed twice? Once for the reflection, and a second time to get the attributes?
Also, do we want the dependency to BetterReflection?

We might get into a bizarre situation where the latest version of BetterReflection already supports attributes reading (but only works for PHP 8), and projects that want attributes in PHP 7 would use an older version of BetterReflection in combination with spiral/attributes.

Perhaps that's all not so bad, but..

## Question / request
I am wondering if spiral wants to support attributes discovery without loading the classes, or what should be the recommended way to achieve this.

One option, as mentioned, would be BetterReflection.

Another option would be to redesign the system a little, so that it would no longer require native `\Reflector` objects (`\ReflectionClass`, `\ReflectionFunction` etc). Instead it could rely on string names of a class, function etc, or on some kind of light-weight handle object, that uniquely identifies the symbol (class, function, parameter etc), without mimicking the full reflection API.

A handle object would also allow to simplify the interface, now we would only need 1 or 2 methods instead of 5 or 10 methods. On the other hand, we would get some switch() or if/else within some of the methods. Of course for BC this would have to be an alternative interface, we cannot just remove the existing ReaderInterface.

## SymbolHandle ideas
Before I discovered `spiral/reflection`, I was experimenting with a custom solution.
Part of this was a `SymbolHandle` that identifies a class, method, parameter, etc.

Basically I ended up with a value object that stores the class name of the reflector class that would be used (e.g. "ReflectionClass", "ReflectionMethod", etc), and the arguments that would be sent to the constructor.

Perhaps this deserves a separate issue.

**Comment by SerafimArts** [comment]
If I understood the task correctly (described in the preface), then the problem here is not in the ReflectionClass\ReflectionEtc..., but in the fact that the attributes are forcibly instantiated (for compatibility with doctrine attributes).

And to solve the problem, it makes sense to select a separate set of methods that implement the return of "prototypes", like this:
```php
interface ReaderInterface
{
    /**
      * Actual method
      * @return iterable<object>
      */
    public function getClassMetadata(\ReflectionClass $class, string $name = null): iterable;

    /**
      * Excpected method
      * @return iterable<ClassAttributeInterface>
      */
    public function getClassMetadataPrototypes(\ReflectionClass $class, string $name = null): iterable;
}
```

Where `ClassAttributeInterface` provides:
```php
interface ClassAttributeInterface
{
    public function exists(): bool;
    public function getName(): string; // attr fqn class string
    public function getArguments(): iterable; // passed arguments list

    public function newInstance(InstantiatorInterface $instantiator = null): object; // can throw "class not found exception"
    public function getReflection(): \ReflectionClass; // same: can throw "class not found exception"
}
```

This way we can, for example, filter out all attributes that have no implementation:
```php
foreach ($reader->getClassMetadataPrototypes($class) as $prototype) {
    if ($prototype->exists()) {
        // do domething
        $attribute = $prototype->new();
    }
}
```

What do you think?

**Comment by donquixote** [comment]
> If I understood the task correctly (described in the preface), then the problem here is not in the ReflectionClass\ReflectionEtc..., but in the fact that the attributes are forcibly instantiated (for compatibility with doctrine attributes).

Actually, I mean both problems:
- Some projects want to avoid loading the class file where the attributes are used.
- Some projects want to avoid loading the class file where the attributes are defined.
- (bonus) Some projects want to avoid loading the class file for an object within the attribute/annotation _arguments_. This only applies with annotations, or in future versions of attributes where objects are allowed in the args.

The main problem would occur if any of those classes have a parent type in an optional 3rd party package.
If the class ifself is not available, we only have to deal with `\ReflectionException`, or `class_exists()` returning FALSE.

However, I now found this:
- Since PHP 7.3, we can catch the Error from extends: https://3v4l.org/mch0d
- Since PHP 7.4 we can catch the Error from implements: https://3v4l.org/0VPev

Interesting, isn't it?

So now I am doubting myself.
I saw this concern in Drupal in the past, but now do we still care?
(asking in Drupal issue queue, https://www.drupal.org/project/drupal/issues/3086628#comment-14240627)

The only remaining concern could be if a class is loaded before autoload is correctly configured. But is this really a concern?

**Comment by donquixote** [comment]
> getClassMetadataPrototypes()

Actually we would not need a new method for this.
A new method would be a BC break for any 3rd parties that already implement the interface. Also it would clutter the interface even more.

Instead, we could simply have an alternative instantiator, which would create a prototype / stub object.
Or another one that would create a `\ReflectionAttribute` object. Perhaps `\ReflectionAttribute` could be added in https://github.com/symfony/polyfill-php80?

Another option would be to use the iterator key for the class name, and the iterator value for the arguments.
Unlike arrays, iterator keys can be repeated.
This prevents the class from being instantiated, but not any objects in the arguments.
This would not work for the single-value method, and would be problematic for BC, so it would live in an alternative reader interface.

**Comment by SerafimArts** [comment]
> Some projects want to avoid loading the class file where the attributes are used.

For what? o___0 I don’t remember a single situation when this would be required. The syntax of attributes is fully compatible with any PHP version (in the case of a one-line declaration), so there shouldn't be any problems associated with them. Therefore, such classes can be safely loaded into memory.

> The main problem would occur if any of those classes have a parent type in an optional 3rd party package.
If the class ifself is not available, we only have to deal with \ReflectionException, or class_exists() returning FALSE.

Ah, that's the problem, now I get it! It is a rather controversial implementation of the package that it uses non-existent classes. However, the team and I will need to think about this case, but it seems to me that this is a rather redundant functionality that violates the behavior of native PHP 8 attributes.

> Or another one that would create a \ReflectionAttribute object. Perhaps \ReflectionAttribute could be added in https://github.com/symfony/polyfill-php80?

You forget that besides attributes, there are also annotations and other types of metadata (config files for example). Although the package is initially called "attributes", I want to stick to a more neutral name when naming methods.

> Actually we would not need a new method for this.
> A new method would be a BC break for any 3rd parties that already implement the interface. Also it would clutter the interface even more.

Yep. Agree.

> Another option would be to use the iterator key for the class name, and the iterator value for the arguments.

We will not be able to expand such an API in the future, adding there any more information, in addition to the class and arguments, as opposed to the object. So it seems to me that the option with a "prototype object" is more promising.

**Comment by donquixote** [comment]
Let's wait if we get more info from the Drupal issue. Perhaps nobody remembers why they used StaticReflectionParser :)
Some links:
https://github.com/doctrine/reflection contains StaticReflectionParser, but is now abandoned.
https://github.com/roave/BetterReflection also advertises with "You can reflect on classes that are not already loaded, without loading them", but I am not sure what's the main use case.
https://git.drupalcode.org/project/drupal/-/blob/a0a22f8d41d7f5926378bcef00b5bb9256ba5fc2/core/lib/Drupal/Component/Annotation/Doctrine/StaticReflectionParser.php the class from doctrine was moved into Drupal, to avoid depending on an abandoned package.
https://git.drupalcode.org/project/drupal/-/blob/a0a22f8d41d7f5926378bcef00b5bb9256ba5fc2/core/lib/Drupal/Component/Annotation/Plugin/Discovery/AnnotatedClassDiscovery.php#L142 - here is where the file is used in a discovery process.

**Comment by donquixote** [comment]
Another problem could be if we run discovery in unknown 3rd party packages, and some of those contain junk classes that would blow up if included, but that can be safely skipped with a parser-based discovery mechanism. But, is this worth the trouble?

**Comment by donquixote** [comment]
> Where `ClassMetadataInterface` provides:

I would have another idea.

```php
interface MetadataFinderInterface {

  /**
   * @param \ReflectionClass $class
   *
   * @return \Donquixote\Ock\MetadataList\MetadataListInterface
   */
  public function classGetMetadata(\ReflectionClass $class): MetadataListInterface;
  [...]
}

interface MetadataListInterface extends \Countable {

  /**
   * @param string|null $type
   *   Class or interface name to filter the results.
   *
   * @return int
   */
  public function count(string $type = NULL): int;

  /**
   * @template T
   *
   * @param class-string<T> $type
   *   Class or interface name to filter the results.
   *
   * @return T[]
   */
  public function getInstances(string $type = NULL): array;

}
```

I think this would be perfect for typical "discovery" use cases.

E.g. imagine a system where you can annotate classes and static methods to be registered as services, or as controllers with routes.

We can now use the visitor pattern like so:

```php
interface FactoryVisitorInterface {

  /**
   * Visits an annotated class.
   *
   * @param \ReflectionClass $class
   * @param \Donquixote\Ock\MetadataList\MetadataListInterface $metadata
   *
   * @throws \Exception
   */
  public function visitAnnotatedClass(\ReflectionClass $class, MetadataListInterface $metadata): void;

  /**
   * Visits an annotated method.
   *
   * @param \ReflectionMethod $method
   * @param \Donquixote\Ock\MetadataList\MetadataListInterface $metadata
   *
   * @throws \Exception
   */
  public function visitAnnotatedMethod(\ReflectionMethod $method, MetadataListInterface $metadata): void;

}
```

**Comment by donquixote** [comment]
> This way we can, for example, filter out all attributes and annotations that have no implementation:

We could add additional methods or parameters in `MetadataListInterface` to deal with undefined classes.


```php
interface MetadataListInterface extends \Countable {
  [..]

  /**
   * @template T
   *
   * @param class-string<T> $type
   *   Class or interface name to filter the results.
   *   Also allows sub-types.
   *
   * @return T[]
   *   Instances where the class exists.
   *   If $type is provided: Instances where (* instanceof $type) would return TRUE.
   */
  public function getInstancesIfClassExists(string $type = NULL): array;
  [..]

  /**
   * @template T
   *
   * @param class-string<T> $type
   *   Class or interface name to filter the results.
   *   Also allows sub-types.
   *
   * @return T[]
   *   If $type is provided: Instances where (* instanceof $type) would return TRUE.
   *
   * @throws \Exception
   *   One of the classes does not exist.
   */
  public function forceGetInstances(string $type = NULL): array;

}
```

I like to have dedicated methods if we have different contracts: One that is allowed to throw an exception, another that is not.
Otherwise, the IDE (and Psalm) will always complain about missing `@throws` tags.

Or alternatively we could have filter methods that return another MetadataListInterface object:

```php
interface MetadataListInterface extends \Countable {
  [..]

  /**
   * @param string $name
   *   Class name to filter by.
   *   Only allows instances where get_class($obj) === $name.
   *
   * @return self
   *   Filtered list.
   */
  public function filterByName(string $name): self;

  /**
   * @param string $type
   *   Class name to filter by.
   *   Only allows instances where $obj instanceof $name.
   *
   * @return self
   *   Filtered list.
   *
   * @todo What about undefined classes?
   */
  public function filterByType(string $type): self;
}
```

Another alternative to `@throws` tags would be a by-reference parameter to collect unknown class names.

```php
  /**
   * @param string $type
   *   Class name to filter by.
   *   Only allows instances where $obj instanceof $name.
   * @param string[]|null $unknownClasses
   *    Will be filled with unknown class names.
   *    Indices are the original positions.
   *
   * @return self
   *   Filtered list.
   *
   * @todo What about undefined classes?
   */
  public function filterByType(string $type, ?array &$unknownClasses): self;

  public function getInstances(string $type, ?array &$unknownClasses): array;
```

I don't think we really need access to the arguments arrays. Most discovery tools won't care about it.
Also considering the case where the arguments themselves might contain `new ..` statements with non-existing classes.
